### DISCUSSION TITLE
Abstract out the socket connection of @redux-devtools/app to allow other communication methods

### DISCUSSION THREAD

**Discussion by matt-oakes** [reduxjs/redux-devtools#1653] [Ideas]
Firstly, thank you for your work on the debugger. It's an invaluable tool for working with Redux.

**Problem**
I mainly use Redux in non-web scenarios, such as in React Native (through Expo) or Electron apps. Currently it is a bit complicated to use the Redux Devtools as you can't just use the extension. The main way to do it currently is using the "remote" option. This works fine, but it's a bit complicated to set up and brittle if you need to also run a local server through the CLI.

**Possible usecases**
My ideal solution would be to use `@redux-devtools/app` with a different way to connect it to the store instead of the socket cluster library.

In the case of Expo, this could be using their [Dev Tools Plugin](https://docs.expo.dev/debugging/devtools-plugins/) infrastructure which is essentially a web socket connection that Expo manages to make the connection between the Expo app (which could be running on a simulator or physical device) and the dev tools web page (which would be some sort of customised `@redux-devtools/app`). There's an [open discussion](https://github.com/expo/dev-plugins/issues/20) about adding support for this.

For Electron this could be using IPC to communicate between the main process (in my case this is where my Redux store lives) and a renderer (which could be displaying the Redux Devtools app).

**Possible solution**
This would be possible now using the lower level packages, but I beleive that you would end up duplicating a lot of the code in `@redux-devtools/app`.

It would be great if there was a way to swap out just the `socketcluster` part of `@redux-devtools/app` and replace it with something specific to each environment. I beleive this would be main replacing the API middleware as that seems to be the only part that directly deals with the socket connection.

Potentially this could be the creation of a `@redux-devtools/app-core` package which allows you to pass in the API middleware you want to use and `@redux-devtools/app` would be changed to just use this core with the middleware for `socketcluster`, as it does now.

My question is, does something like this make sense? I would be happy to look at the implementation of this, but as it's a fairly big change I wanted to start a discussion before starting to work on it. I don't really want to maintain a fork of the devtools so having something you would be happy to merge in would be best.

Thanks again!

**Comment by markerikson** [comment]
I'm not the primary maintainer here, but that does sound like a reasonable suggestion, yeah!

**Reply by matt-oakes** [reply]
That's great to hear. I have some spare time next week so I will begin to take a look at this and put together a draft PR for the maintainer to look through.

The goal will be to make as few changes as possible and not make any breaking changes to the `@redux-devtools/app` package so it can be used as it currently is.

**Reply by Methuselah96** [reply]
Thanks for your interest in working on this! I appreciate you creating a discussion first.

The two things I'll note:
 - I'm not particularly tied to socketcluseter. The choice to use socketcluster was inherited and it seems to not be all that well maintained or popular, so at some point I'm interested in moving to a different library. This might not affect anything and is probably too large of a task to bite off at this time, but I figured it could be helpful context.
 - I'm pretty bad at reviewing/merging PRs, especially when they're quite large, usually because there are some small parts of that seem questionable to me and it takes a lot of energy for me to review large PRs. I say that just to encourage you to make the PR as small and straightforward as possible.

**Reply by matt-oakes** [reply]
Thanks for the quick reply.

Keeping the PR as small as possible makes sense. I beleive there will be a fairly large diff of changes which are inherent to creating a new `@redux-devtools/app-core` package and moving the majority of the current `@redux-devtools/app` code over to it.

If you woud prefer, I could start with a PR which just makes this change and makes `@redux-devtools/app` into essentially a shim which just re-exports the new `app-core` package. I can then follow that up with a PR which actually adds the abstractions and moves the use of `socketcluster` back into just `@redux-devtools/app`.

This would give you two smaller, more focused reviews to do, rather than one larger one. It's up to you which you would prefer. I am happy to submit it that way, or just as a single PR.

**Reply by Methuselah96** [reply]
I think the idea is straightforward enough that it should be fine either way. Thanks for taking this on!

**Comment by matt-oakes** [comment]
I have opened a PR with the changes discussed https://github.com/reduxjs/redux-devtools/pull/1655

I will close this discussion now as any comments can be left on the PR itself.

**Comment by fai8146767** [comment]
****
